arrow-json/src/reader/struct_array.rs

```diff
@@ -216,46 +238,90 @@ impl ArrayDecoder for StructArrayDecoder {
             match fields.iter().position(|x| x.name() == field_name) {
                 Some(field_idx) => {
                     let child_pos = cur_idx + 1;
-                    if !self.decoders[field_idx].validate_row(tape, child_pos) {
-                        return false;
+
+                    if let Err(mut child_errors) =
+                        self.decoders[field_idx].validate_row(tape, child_pos, row_idx)
+                    {
+                        // Add field name to child errors that don't have one
+                        for error in &mut child_errors {
+                            if error.field_name.is_none() {
+                                error.field_name = Some(field_name);
+                            }
+                        }
+                        return Err(child_errors);
                     }
+
                     validated_fields[field_idx] = true;
                 }
                 None => {
                     if self.strict_mode {
-                        return false;
+                        // Custom field_name - can't use helper
+                        return Err(vec![ErrorMarker {
+                            row_index: row_idx,
+                            tape_pos: cur_idx,
+                            field_name: Some(field_name),
+                            array_indices: Vec::new(),
+                            error_kind: FailureKind::TypeMismatch,
+                            expected_type: Arc::clone(&self.data_type),
+                        }]);
                     }
                 }
             }
 
-            // Advance to next field
             cur_idx = match tape.next(cur_idx + 1, "field value") {
                 Ok(i) => i,
                 Err(_) => {
-                    return false;
+                    return ErrorMarker::err(
+                        row_idx,
+                        cur_idx,
+                        FailureKind::TypeMismatch,
+                        Arc::clone(&self.data_type),
+                    );
                 }
-            }
+            };
         }
 
-        validated_fields
+        // Early exit on happy path - no missing field errors
+        let all_valid = validated_fields
             .iter()
             .zip(fields)
-            .all(|(validated, field)| {
-                if !validated && !field.is_nullable() {
-                    return false;
-                }
-                true
-            })
+            .all(|(validated, field)| *validated || field.is_nullable());
+
+        if all_valid {
+            return Ok(());
+        }
+
+        // Error path: collect missing field errors
+        let mut missing_errors = Vec::new();
+        for (validated, field) in validated_fields.iter().zip(fields) {
+            if !validated && !field.is_nullable() {
+                missing_errors.push(ErrorMarker {
+                    row_index: row_idx,
+                    tape_pos: u32::MAX,
```

Comment:
Can you explain why we're using u32::MAX here?

Reply:
`tape_pos` is an optional now.